File: data-structures/linear/linked-list/circular-doubly-linked-list/CircularDoublyLinkedList.java
class Node {
  int data;
  Node next;
  Node prev;

  Node(int value) {
    this.data = value;
    this.next = this.prev = null;
  }
}

class DoublyCircularLinkedList {
  private Node head;
  private Node tail;

  DoublyCircularLinkedList() {
    this.head = null;
    this.tail = null;
  }

  // ------------- Core Operations -------------

  boolean isEmpty() {
    return head == null;
  }

  void TraverseForward() {
    if (head == null) {
      System.out.println("Empty");
      return;
    }
    Node temp = head;
    do {
      System.out.print(temp.data + " <-> ");
      temp = temp.next;
    } while (temp != head);
    System.out.println("(back to head)");
  }

  void TraverseBackward() {
    if (head == null) {
      System.out.println("Empty");
      return;
    }
    Node temp = tail;
    do {
      System.out.print(temp.data + " <-> ");
      temp = temp.prev;
    } while (temp != tail);
    System.out.println("(back to tail)");
  }

  void Search(int searchData) {
    if (head == null) {
      System.out.println("Empty");
      return;
    }
    int pos = 0;
    Node temp = head;
    do {
      if (temp.data == searchData) {
        System.out.println("Data " + temp.data + " found at position " + pos);
        return;
      }
      pos++;
      temp = temp.next;
    } while (temp != head);
    System.out.println("Data " + searchData + " not found");
  }

  int front() {
    if (head == null) {
      System.out.println("List empty");
      return -1;
    }
    return head.data;
  }

  int back() {
    if (tail == null) {
      System.out.println("List empty");
      return -1;
    }
    return tail.data;
  }

  int size() {
    if (head == null)
      return 0;
    int count = 0;
    Node temp = head;
    do {
      count++;
      temp = temp.next;
    } while (temp != head);
    return count;
  }

  // ------------- Insertions -------------

  void insertFirst(int value) {
    Node newNode = new Node(value);
    if (head == null) {
      head = tail = newNode;
      newNode.next = newNode.prev = newNode;
    } else {
      newNode.next = head;
      newNode.prev = tail;
      tail.next = newNode;
      head.prev = newNode;
      head = newNode;
    }
  }

  void insertLast(int value) {
    Node newNode = new Node(value);
    if (head == null) {
      head = tail = newNode;
      newNode.next = newNode.prev = newNode;
    } else {
      newNode.prev = tail;
      newNode.next = head;
      tail.next = newNode;
      head.prev = newNode;
      tail = newNode;
    }
  }

  void insertAt(int position, int value) {
    if (position == 0) {
      insertFirst(value);
      return;
    }

    Node temp = head;
    int count = 0;
    while (count < position - 1 && temp.next != head) {
      temp = temp.next;
      count++;
    }

    if (count != position - 1) {
      System.out.println("Position out of range");
      return;
    }

    Node newNode = new Node(value);
    newNode.next = temp.next;
    newNode.prev = temp;
    temp.next.prev = newNode;
    temp.next = newNode;

    if (temp == tail) {
      tail = newNode;
    }
  }

  // ------------- Deletions -------------

  void deleteFirst() {
    if (head == null)
      return;

    if (head == tail) {
      System.out.println("Deleted value : " + head.data);
      head = tail = null;
      return;
    }
    System.out.println("Deleted value : " + head.data);
    head = head.next;
    head.prev = tail;
    tail.next = head;
  }

  void deleteLast() {
    if (head == null)
      return;

    if (head == tail) {
      System.out.println("Deleted value : " + head.data);
      head = tail = null;
      return;
    }
    System.out.println("Deleted value : " + tail.data);
    tail = tail.prev;
    tail.next = head;
    head.prev = tail;
  }

  void deleteAt(int position) {
    if (head == null)
      return;
    if (position == 0) {
      deleteFirst();
      return;
    }

    Node temp = head;
    int count = 0;
    while (count < position - 1 && temp.next != head) {
      temp = temp.next;
      count++;
    }

    if (count != position - 1 || temp.next == head) {
      System.out.println("Position out of range");
      return;
    }

    Node toDelete = temp.next;
    System.out.println("Deleted: " + toDelete.data);
    temp.next = toDelete.next;
    toDelete.next.prev = temp;

    if (toDelete == tail) {
      tail = temp;
    }
  }

  void deleteKey(int deleteData) {
    if (head == null)
      return;

    Node temp = head;
    do {
      if (temp.data == deleteData) {
        if (temp == head && temp == tail) {
          System.out.println("Deleted value: " + temp.data);
          head = tail = null;
          return;
        } else if (temp == head) {
          deleteFirst();
          return;
        } else if (temp == tail) {
          deleteLast();
          return;
        } else {
          temp.prev.next = temp.next;
          temp.next.prev = temp.prev;
          System.out.println("Deleted value: " + temp.data);
          return;
        }
      }
      temp = temp.next;
    } while (temp != head);

    System.out.println("Node with given data: " + deleteData + " not found");
  }

  // ------------- Utilities -------------
  // 😎💀📔
  void update(int updateVal, int newData) {
    if (head == null) {
      System.out.println("Empty List");
      return;
    }
    Node temp = head;
    do {
      if (temp.data == updateVal) {
        temp.data = newData;
        System.out.println("Node updated: " + updateVal + " -> " + newData);
        return;
      }
      temp = temp.next;
    } while (temp != head);
    System.out.println("Node with given data: " + updateVal + " not found");
  }

  void reverse() {
    if (head == null) {
      System.out.println("Empty List");
      return;
    }
    Node current = head, temp = null;
    do {
      temp = current.prev;
      current.prev = current.next;
      current.next = temp;
      current = current.prev;
    } while (current != head);

    temp = head;
    head = tail;
    tail = temp;
  }
}

public class CircularDoublyLinkedList {
  public static void main(String[] args) {
    DoublyCircularLinkedList list = new DoublyCircularLinkedList();

    list.insertLast(10);
    list.insertLast(20);
    list.insertLast(30);
    list.TraverseForward();

    list.insertFirst(5);
    list.TraverseForward();

    list.insertAt(2, 15);
    list.TraverseForward();

    System.out.println("Front: " + list.front());
    System.out.println("Back: " + list.back());
    System.out.println("Size: " + list.size());

    list.deleteFirst();
    list.TraverseForward();

    list.deleteLast();
    list.TraverseForward();

    list.deleteAt(1);
    list.TraverseForward();

    list.deleteKey(15);
    list.TraverseForward();

    list.update(20, 200);
    list.TraverseForward();

    list.reverse();
    list.TraverseForward();
    list.TraverseBackward();
  }
}

File: data-structures/linear/linked-list/circular-doubly-linked-list/CircularDoublyLinkedList.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insertFirst(self, val):
        new_node = Node(val)
        if self.head is None:
            new_node.next = new_node
            new_node.prev = new_node
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            new_node.prev = self.tail
            self.head.prev = new_node
            self.tail.next = new_node
            self.head = new_node

    def insertLast(self, val):
        new_node = Node(val)
        if self.head is None:
            new_node.next = new_node
            new_node.prev = new_node
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            new_node.next = self.head
            self.tail.next = new_node
            self.head.prev = new_node
            self.tail = new_node

    def deleteFirst(self):
        if self.head is None:
            return
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = self.tail
            self.tail.next = self.head

    def deleteLast(self):
        if self.head is None:
            return
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = self.head
            self.head.prev = self.tail

    def insertAfter(self, u, v):
        if self.head is None:
            print(f"Value {u} not found in list.")
            return
        curr = self.head
        while True:
            if curr.val == u:
                new_node = Node(v)
                new_node.next = curr.next
                new_node.prev = curr
                curr.next.prev = new_node
                curr.next = new_node
                if curr == self.tail:
                    self.tail = new_node
                return
            curr = curr.next
            if curr == self.head:
                break
        print(f"Value {u} not found in list.")

    def linear_search(self, target):
        if self.head is None:
            return -1
        index = 0
        curr = self.head
        while True:
            if curr.val == target:
                return index
            curr = curr.next
            index += 1
            if curr == self.head:
                break
        return -1

    def listSize(self):
        if self.head is None:
            return 0
        size = 1
        curr = self.head.next
        while curr != self.head:
            size += 1
            curr = curr.next
        return size

    def isEmpty(self):
        return self.head is None

    def printList(self):
        if self.head is None:
            print("Empty list")
            return
        curr = self.head
        while True:
            print(curr.val, end=" <-> ")
            curr = curr.next
            if curr == self.head:
                break
        print("(back to head)")
